src/ripple/app/ledger/OpenLedger.h

```diff
@@ -0,0 +1,268 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_LEDGER_OPENLEDGER_H_INCLUDED
+#define RIPPLE_APP_LEDGER_OPENLEDGER_H_INCLUDED
+
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/ledger/CachedSLEs.h>
+#include <ripple/ledger/OpenView.h>
+#include <ripple/app/misc/CanonicalTXSet.h>
+#include <ripple/app/misc/IHashRouter.h>
+#include <ripple/basics/Log.h>
+#include <ripple/basics/UnorderedContainers.h>
+#include <ripple/core/Config.h>
+#include <beast/container/aged_unordered_map.h>
+#include <beast/utility/Journal.h>
+#include <cassert>
+#include <mutex>
+
+namespace ripple {
+
+// How many total extra passes we make
+// We must ensure we make at least one non-retriable pass
+#define LEDGER_TOTAL_PASSES 3
+
+// How many extra retry passes we
+// make if the previous retry pass made changes
+#define LEDGER_RETRY_PASSES 1
+
+using OrderedTxs = CanonicalTXSet;
+
+//------------------------------------------------------------------------------
+
+/** Represents the open ledger. */
+class OpenLedger
+{
+private:
+    beast::Journal j_;
+    CachedSLEs& cache_;
+    Config const& config_;
+    std::mutex mutable modify_mutex_;
```

Comment:
There's not much that can be added that the code doesn't already explain, one serializes modifications and the other serializes access to the current view.